components/service/glean/src/main/java/mozilla/components/service/glean/storages/CustomDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,384 @@
+     * bucket where the sample is bigger than or equal to the bucket limit.
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun findBucket(sample: Long): Long {
+        var under = 0
+        var over = bucketCount
+        var mid: Int
+
+        do {
+            mid = under + (over - under) / 2
+            if (mid == under) {
+                break
+            }
+
+            if (buckets[mid] <= sample) {
+                under = mid
+            } else {
+                over = mid
+            }
+        } while (true)
+
+        return buckets[mid]
+    }
+
+    /**
+     * Accumulates a sample to the correct bucket.
+     * If a value doesn't exist for this bucket yet, one is created.
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun accumulate(sample: Long) {
+        val limit = findBucket(sample)
+        values[limit] = (values[limit] ?: 0) + 1
+        sum += sample
+    }
+
+    /**
+     * Helper function to build the [CustomDistributionData] into a JSONObject for serialization
+     * purposes.
+     */
+    internal fun toJsonObject(): JSONObject {
+        return JSONObject(mapOf(
+            "category" to category,
+            "name" to name,
+            "bucket_count" to bucketCount,
+            "range" to JSONArray(arrayOf(rangeMin, rangeMax)),
+            "histogram_type" to histogramType.toString().toLowerCase(),
+            "values" to values.mapKeys { "${it.key}" },
+            "sum" to sum
+        ))
+    }
+
+    /**
+     * Helper function to generate the list of linear bucket min values used when accumulating
+     * to the correct buckets.
+     *
+     * @return List containing the bucket limits
+     */
+    @Suppress("MagicNumber")
+    private fun getBucketsLinear(): List<Long> {
+        // Written to match the bucket generation on legacy desktop telemetry:
+        //   https://searchfox.org/mozilla-central/source/toolkit/components/telemetry/build_scripts/mozparsers/parse_histograms.py#65
+
+        val result: MutableList<Long> = mutableListOf()
+        result.add(0)
+
+        val dmin = rangeMin.toDouble()
+        val dmax = rangeMax.toDouble()
+
+        for (i in (1 until bucketCount)) {
+            val linearRange = (dmin * (bucketCount - 1 - i) + dmax * (i - 1)) / (bucketCount - 2)
+            result.add((linearRange + 0.5).toLong())
+        }
+
+        return result
+    }
+
+    /**
+     * Helper function to generate the list of exponential bucket min values used when accumulating
+     * to the correct buckets.
+     *
+     * @return List containing the bucket limits
+     */
+    private fun getBucketsExponential(): List<Long> {
+        // Written to match the bucket generation on legacy desktop telemetry:
+        //   https://searchfox.org/mozilla-central/source/toolkit/components/telemetry/build_scripts/mozparsers/parse_histograms.py#75
```

Comment:
Would you mind changing this to the permalink as well? https://searchfox.org/mozilla-central/rev/e0b0c38ee83f99d3cf868bad525ace4a395039f1/toolkit/components/telemetry/build_scripts/mozparsers/parse_histograms.py#75